Include Go version, platform, and other build info in version string

ixdy opened the pull request:
Tracking down the root cause of #17524 probably would've been easier if we'd more readily exposed the Go version used to build Kubernetes, so this PR adds it to the version string exported.

This is fairly similar to the build information contained in Linux's version string or the command `docker version`.

Note that this will make [`MatchesServerVersion`](https://github.com/kubernetes/kubernetes/blob/36dc6eab3437063147d6daac76add2258eb37561/pkg/client/unversioned/helper.go#L151) fail if the client/server were built at the same git hash but with a different version of Go. I'm not sure if this is something to worry about.

I don't think this is likely to break anything, but I'm not entirely confident on that.

(I also considered whether we want to add a build timestamp, like Linux and Docker have, but then we'd definitely want to ignore it in `MatchesServerVersion`, so I left that out for now.)

ixdy committed:
Include Go version, platform, and other build info in version string

Additionally update MatchesServerVersion to only check GitVersion,
GitCommit, and GitTreeState.

ixdy commented:
@k8s-bot test this please

ixdy commented:
@k8s-bot e2e test this please

luxas commented:
Will this make `v1.2`?

ixdy commented:
@luxas it's been unclear to me if anyone wants it...

ixdy commented:
@k8s-bot test this please, issue #IGNORE (Jenkins got stuck at some point?)

bgrant0607 commented:
@lavalamp @liggitt @deads2k Can any of you think of a problem this would cause? It seems like the git hashes would mismatch any time the go versions differed.

liggitt commented:
wouldn't `kubectl --match-server-version` complain about a client built with a different go version than the server? that seems less than ideal

deads2k commented:
> wouldn't kubectl --match-server-version complain about a client built with a different go version than the server? that seems less than ideal

An argument to fix `MatchesServerVersion`, not one to avoid adding this information.  It's a little esoteric, but as long as `MatchesServerVersion` is updated to skip trying to match on go compiler level I don't see it as a problem.

smarterclayton commented:
Fortunately matchesServerVersion doesn't have to be backwards compatible

liggitt commented:
> Fortunately matchesServerVersion doesn't have to be backwards compatible

heh, true

lavalamp commented:
I think the go version skew between client & server is a) hard to get in practice b) worth noting when it happens. So I think this is probably net good. LGTM

ixdy commented:
So what's the consensus here? Do we want to update `MatchesServerVersion` to ignore the go version? Or is this fine as-is?

deads2k commented:
> So what's the consensus here? Do we want to update MatchesServerVersion to ignore the go version? Or is this fine as-is?

I'd like to ignore the go version and base the decision on whether server and client were built from the same code.

lavalamp commented:
I vote for it being fine as-is. I think client & server built with
different versions of go should be considered different. The only way you
can convince me otherwise is if cross-platform client/server combos would
trip over this.

On Thu, Mar 3, 2016 at 5:01 AM, David Eads <notifications@github.com> wrote:

> So what's the consensus here? Do we want to update MatchesServerVersion to
> ignore the go version? Or is this fine as-is?
>
> I'd like to ignore the go version and base the decision on whether server
> and client were built from the same code.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19905#issuecomment-191752313>
> .
>

ixdy commented:
The way it is is certainly less work for me. :)

Otherwise I will probably need to clear out the client and server Go versions in `MatchesServerVersion`, since I'd prefer to keep using `reflect.DeepEqual` instead of enumerating all fields.

(It's also possible there's code elsewhere that compares versions, which would also need to account for Go versions? I can do another pass to see. It's kinda unfortunate that `MatchesServerVersion` is in pkg/client/unversioned/helper.go instead of pkg/version, where it'd probably be more appropriate, but that's also a more in-depth refactoring effort.)

lavalamp commented:
If the platform isn't in the go version then I'm good with it the way it
is. @deads2k why do you think it will be problematic?

On Thu, Mar 3, 2016 at 4:17 PM, Jeff Grafton <notifications@github.com>
wrote:

> The way it is is certainly less work for me. :)
>
> Otherwise I will probably need to clear out the client and server Go
> versions in MatchesServerVersion, since I'd prefer to keep using
> reflect.DeepEqual instead of enumerating all fields.
>
> (It's also possible there's code elsewhere that compares versions, which
> would also need to account for Go versions? I can do another pass to see.
> It's kinda unfortunate that MatchesServerVersion is in
> pkg/client/unversioned/helper.go instead of pkg/version, where it'd
> probably be more appropriate, but that's also a more in-depth refactoring
> effort.)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19905#issuecomment-192032871>
> .
>

ixdy commented:
@lavalamp what the version looks like now (from the most recent e2e run of this PR):
```
2016/01/21 20:13:42 e2e.go:303: Running: 'kubectl version --match-server-version=false'
Client Version: version.Info{Major:"1", Minor:"2+", GitVersion:"v1.2.0-alpha.6.361+832d81820d7497", GitCommit:"832d81820d749749409495d7536274ee76e706ce", GitTreeState:"clean", GoVersion:"go1.4.2"}
Server Version: version.Info{Major:"1", Minor:"2+", GitVersion:"v1.2.0-alpha.6.361+832d81820d7497", GitCommit:"832d81820d749749409495d7536274ee76e706ce", GitTreeState:"clean", GoVersion:"go1.4.2"}
```

So no platform info.

ixdy commented:
@k8s-bot test this please, issue #22435

luxas commented:
It might be good with platform in there too. Now, we already support ARM with #19769, and it can and will happen for instance that one controls an ARM machine via `amd64` kubectl. When we have `go1.6` we'll support even more arches: `arm64` and `ppc64le`. So to make it easier to debug and such we should take platform into account. Docker does this already.

deads2k commented:
> If the platform isn't in the go version then I'm good with it the way it
> is. @deads2k why do you think it will be problematic?

The primary utility of the check (in my experience) has been to ensure that the API kinds are identical between client and server.  That information makes it safe to do things like client-side conversions and client-side edits for UDPATE without fear of dropping fields.  The source of that information is the git commit, the particular compiler used doesn't enter into it so I don't see a reason to restrict it.

Adding "platform" to the comparison would actively break the utility and power of the feature, rendering it mostly useless since there's not strong reason to believe that clients and servers run the same OS.

deads2k commented:
I guess I would ask what you use the `--matches-server-version` flag for that would work better by forcing compilers to match.

luxas commented:
I would be fine with including `goVersion` and `platform` in `kubectl version`, but let `matchesServerVersion` focus on just the commit for the reason @deads2k pointed out

lavalamp commented:
I agree the --matches-server-version flag should *not* check the platform.

I *do* think different compiler versions are a (very very rare) source of
incompatibilities. On a scale of 1-10, I'm about a 2 for "yes, check it",
mainly because I can't think of a plausible scenario where it's annoying to
check that. @deads2k can you think of such a scenario?

On Fri, Mar 4, 2016 at 9:05 AM, Lucas Käldström <notifications@github.com>
wrote:

> I would be fine with including goVersion and platform in kubectl version,
> but let matchesServerVersion focus on just the commit for the reason
> @deads2k <https://github.com/deads2k> pointed out
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19905#issuecomment-192359592>
> .
>

deads2k commented:
> On a scale of 1-10, I'm about a 2 for "yes, check it",
> mainly because I can't think of a plausible scenario where it's annoying to
> check that. @deads2k can you think of such a scenario?

No, but I'd err on the other side.  Is it really any harder to update https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/helper.go#L106 to `if serverVersion.GitCommit == clientVersion.GitCommit && serverVersion.GitTreeState == clientVersion.GitTreeState`?

I guess its changeable later if we need to change it, so there's no reason to fight over it.

lavalamp commented:
Like I said, I'm a 2 so I don't care too much.

On Mon, Mar 7, 2016 at 5:35 AM, David Eads <notifications@github.com> wrote:

> On a scale of 1-10, I'm about a 2 for "yes, check it",
> mainly because I can't think of a plausible scenario where it's annoying to
> check that. @deads2k <https://github.com/deads2k> can you think of such a
> scenario?
>
> No, but I'd err on the other side. Is it really any harder to update
> https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/helper.go#L106
> to if serverVersion.GitCommit == clientVersion.GitCommit &&
> serverVersion.GitTreeState == clientVersion.GitTreeState?
>
> I guess its changeable later if we need to change it, so there's no reason
> to fight over it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19905#issuecomment-193250287>
> .
>

ixdy commented:
Per comments, I updated `MatchesServerVersion` to only check `GitVersion`, `GitCommit`, and `GitTreeState`. (The latter two are probably redundant, but do we care?)

I also added `Platform` per suggestion of @luxas, and also `Compiler` and `BuildDate`.

luxas commented on pkg/version/version.go:
```diff
@@ -52,12 +65,12 @@ func init() {
+			Help: "A metric with a constant '1' value labeled by major, minor, git version, git commit, git tree state, and Go version from which Kubernetes was built.",
 		},
-		[]string{"major", "minor", "gitVersion", "gitCommit", "gitTreeState"},
+		[]string{"major", "minor", "gitVersion", "gitCommit", "gitTreeState", "goVersion"},
```
Shouldn't `Compiler`, `BuildDate` and `Platform` be here also? Or am I missing something...

ixdy replied:
Good catch! Done.

ixdy commented:
Fixed the panic - forgot to update half of the prometheus stuff.

ixdy commented:
It seems go1.5.3 doesn't like spaces in the ldflag value, so I switched to using the ISO8601 format for the build date.

luxas commented:
@ixdy Please squash and maybe @lavalamp could LGTM so we could get this in. 
Will be valuable to have merged

lavalamp commented:
lgtm

lavalamp commented:
apply the label once you squash.

ixdy commented:
Sorry for the noise. Interactive rebased a bit too far, but fixed now.

Thanks for review.
